### PR TITLE
Iter force sequential

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

Introduce strict sequential mode for iterators where one iterator passes every particle even when started in a parallel section.

- [x] Make IteratorBehavior a bitmask
- [x] remove all internal default values

## Related Pull Requests

- merges into #580

## Resolved Issues

- [x] fixes #579 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] ParticleIteratorTest::testForceSequential
- [x] RegionParticleIteratorTest::testForceSequential

### CODE REVIEW

File: src/autopas/iterators/ParticleIterator.h [hunk 1]
```diff
@@ -57,18 +57,20 @@ class ParticleIterator : public ParticleIteratorInterfaceImpl<Particle, modifiab
...
         _iteratorAcrossCells(cont->begin()),
         _iteratorWithinOneCell(cont->begin()->begin()),
         _flagManager(flagManager),
         _behavior(behavior),
+        _forceSequential(forceSequential),
         _additionalParticleVectorToIterateState(additionalVectorsToIterate
                                                     ? AdditionalParticleVectorToIterateState::notStarted
                                                     : AdditionalParticleVectorToIterateState::ignore),
-        _additionalVectorIndex(autopas_get_thread_num()),
+        _additionalVectorIndex(forceSequential ? autopas_get_thread_num() : 0),
```
Comment (SteffenSeckler) [inline src/autopas/iterators/ParticleIterator.h:73]:
```suggestion
        _additionalVectorIndex(forceSequential ? 0 : autopas_get_thread_num()),
```

Comment (SteffenSeckler) [reply src/autopas/iterators/ParticleIterator.h:73]: do you test the case, where you add particles with more threads than with which you iterate (to test the correct iteration using the `_additionalVectorIndex`?)

**Review by SteffenSeckler — CHANGES_REQUESTED** [review]
bug: swapped if for the sequential case.
There should be more documentation inside of the iterators telling a maintainer, why the effect of force sequential is as said.

File: src/autopas/iterators/ParticleIterator.h (hunk 1 above)
Comment (FG-TUM) [reply src/autopas/iterators/ParticleIterator.h:73]: Yes this is tested via three different tests in `ParticleIteratorTest`

File: src/autopas/AutoPas.h [hunk 2]
```diff
@@ -219,27 +219,32 @@ class AutoPas {
...
    * for(auto iter = autoPas.begin(); iter.isValid(); ++iter)
-   * @param behavior the behavior of the iterator. You can specify whether to iterate over owned particles, halo
+   * @param behavior The behavior of the iterator. You can specify whether to iterate over owned particles, halo
    * particles, or both.
+   * @param forceSequential Whether to force the iterator to behave as if it is not parallel.
    * @return iterator to the first particle.
    */
-  iterator_t begin(IteratorBehavior behavior = IteratorBehavior::haloAndOwned) {
-    return _logicHandler->begin(behavior);
+  iterator_t begin(IteratorBehavior behavior = IteratorBehavior::haloAndOwned, bool forceSequential = false) {
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/AutoPas.h:227]: or when using strong types, check such as `.isSequential()`, `.isParallel()`, `.iterateOwned()`, `...` would be possible.

**Comment by SteffenSeckler** [comment]
RegionParticleIteratorTest::testForceSequential ?

File: tests/testAutopas/tests/containers/verletListsCellBased/verletListsCells/VerletListsCellsTest.cpp [hunk 3]
```diff
@@ -37,7 +37,8 @@ void applyFunctor(MockFunctor<Particle> &functor, const double cellSizefactor) {
   verletLists.iteratePairwise(&traversal);
 
   std::vector<Particle *> list;
-  for (auto iter = verletLists.begin(); iter.isValid(); ++iter) list.push_back(&*iter);
+  for (auto iter = verletLists.begin(autopas::IteratorBehavior::ownedOrHalo); iter.isValid(); ++iter)
+    list.push_back(&*iter);
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/containers/verletListsCellBased/verletListsCells/VerletListsCellsTest.cpp:41]:
```suggestion
  for (auto iter = verletLists.begin(autopas::IteratorBehavior::ownedOrHalo); iter.isValid(); ++iter) {
    list.push_back(&*iter);
  }
```

**Review by SteffenSeckler — CHANGES_REQUESTED** [review]
RegionParticleIterator test is missing.

File: src/autopas/options/IteratorBehavior.h [hunk 4]
```diff
@@ -58,20 +89,26 @@ class IteratorBehavior : public Option<IteratorBehavior> {
...
-        {IteratorBehavior::haloAndOwned, "haloAndOwned"},
-        {IteratorBehavior::haloOwnedAndDummy, "haloOwnedAndDummy"},
+        {IteratorBehavior::owned, "owned"},
+        {IteratorBehavior::halo, "halo"},
+        {IteratorBehavior::ownedOrHalo, "ownedOrHalo"},
+        {IteratorBehavior::dummy, "dummy"},
+        {IteratorBehavior::ownedOrHaloOrDummy, "ownedOrHaloOrDummy"},
+        {IteratorBehavior::forceSequential, "forceSequential"},
     };
   }
```
Comment (SteffenSeckler) [inline src/autopas/options/IteratorBehavior.h:114]: possibly, it makes sense to add convenience functions to the behavior enum, such as:
```c++
doesIterateOwned();
goesIterateHalo();
```
I'm not happy with the names, though ^^

File: tests/testAutopas/tests/containers/verletListsCellBased/verletListsCells/VerletListsCellsTest.cpp (hunk 3 above)
Comment (FG-TUM) [reply tests/testAutopas/tests/containers/verletListsCellBased/verletListsCells/VerletListsCellsTest.cpp:41]: what is the suggestion here?^^

Comment (SteffenSeckler) [reply tests/testAutopas/tests/containers/verletListsCellBased/verletListsCells/VerletListsCellsTest.cpp:41]: `{}`

File: src/autopas/options/IteratorBehavior.h (hunk 4 above)
Comment (FG-TUM) [reply src/autopas/options/IteratorBehavior.h:114]: I'm a bit against explicitly creating state checker functions for every single value. However, I could imagine a function like:
```
bool contains(IteratorBehavior behavior) {
  return _value & behavior;
}

File: src/autopas/iterators/RegionParticleIterator.h [hunk 5]
```diff
@@ -74,12 +75,12 @@ class RegionParticleIterator : public ParticleIterator<Particle, ParticleCell, m
       return;
     }
 
-    if (behavior != IteratorBehavior::haloAndOwned and flagManager == nullptr) {
+    if ((behavior & IteratorBehavior::ownedOrHalo) and flagManager == nullptr) {
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/iterators/RegionParticleIterator.h:78]: this check is actually redundant as it is already done in the constructor of ParticleIterator

**Review by SteffenSeckler — APPROVED** [review]

**Pull request merged**